lib/key.js

```
@@ -229,7 +229,12 @@ module.exports = function (keyOptions, clientId, secret, self) {
 
 				return reject(err);
 			});
-		});
+		})
+		// handle any uncaught exceptions
+		.catch((err) => {
+			
+			return err.message;
```

This should be `return reject(err);`

---

Shouldn't it be `return Promise.reject(err)`? I think `reject` is outside of the scope..

---

Yes, good catch!

---

I had it this way initially. Unfortunately, the unhandled promise rejection didn't get fixed killing the realtime app.

---

However,  I could either go with  Promise.resolve(err); or just the way it is in order to send a graceful message to the realtime app.

---

This would be a bad idea - the intended effect here is to indicate that an error has occurred and processing can't complete. For any other consumer of the SDK, to receive a properly resolved promise with an object response that does not match expectations would be really bad. This seems to point at something that is not working as expected in the code that is consuming the SDK... especially if you build a unit test to validate this.

---

Can you modify the code in player.js to handle the reject instead of changing the return type of the method?

---

okay, actually it is the player.js in the sdk calling the key.js for authorization before a socket connection to be established. Should it be handled different if it is events we are listening to?
Here is the code in player.js calling the key.js. 

self.connect = (socketEventSubscriber) => {
		co(function *() {
			let
				authHeaders = yield ensureAuthHeaders(),
*************
*************
socket.on('error', (err) => {
				socketEventSubscriber.emit('error', err);
			});

Realtime just waits on the response this event/s receive.

---

maybe add a catch for the yield ensureAuthHeaders call here and then if it catches the error emit the error?

---

if `ensureAuthHeaders` fails, the `co` should bubble the exception as a rejected Promise. However, notice that the `co` function is not being returned... add a `return` statement before `co` and this will probably help.

---

or rewrite a bit...

```
self.connect = (socketEventSubscriber) => (co(function *() {
// teh codez herr
});
```

---

okay, working on a fix. on the player.js file, thank you.
We found an issue with socket-io-client where it doesn't try to reconnect if failed during during authorization. Craig is looking into whether upgrading socket-io-client is a solution.

---

Here is the link: https://stackoverflow.com/questions/24390482/socket-io-failing-authorization-does-not-retry-to-reconnect